machetli/search.py

```diff
@@ -84,28 +83,21 @@ def search(initial_state, successor_generator, evaluator_path, environment=None)
     successor_generator = make_single_successor_generator(successor_generator)
 
     logging.info("Starting search ...")
-
     current_state = initial_state
-    batch_size = environment.batch_size
-    batch_num = 0
-
-    successors = successor_generator.get_successors(current_state)
-    batch = list(islice(successors, batch_size))
-    while batch:
-        try:
-            environment.submit(batch, batch_num, evaluator_path)
-            environment.wait_until_finished()
-            successor = environment.get_improving_successor()
-        except (SubmissionError, TaskError, PollingError):
-            # FIXME: this is not proper error handling yet.
-            successor = None
-
-        if successor:
-            logging.info(successor.change_msg)
-            current_state = successor.state
-            successors = successor_generator.get_successors(current_state)
-        batch_num += 1
-        batch = list(islice(successors, batch_size))
-
-    logging.info("No improving successor found, terminating search.")
-    return current_state
+    while True:
+        successors = successor_generator.get_successors(current_state)
+        for batch in batched(successors, environment.batch_size):
+            try:
+                environment.submit(batch, evaluator_path)
+                environment.wait_until_finished()
+                successor = environment.get_improving_successor()
+            except (SubmissionError, TaskError, PollingError):
+                # FIXME: this is not proper error handling yet.
+                successor = None
+            if successor:
+                logging.info(successor.change_msg)
+                current_state = successor.state
+                break
+        else:
```

Comment:
This looks a bit weird to me, because the `else` case is on a different indentation level as the `if` statement. Is this some kind of Python magic I'm not (yet) familiar with? I interpret the functionality here as "loop until successor found, if never successor found do `else`", does that make sense? This is also what I think was happening before, so I think the refactoring is fine. I just wanted to make sure I understand Python correctly.

Reply:
I see the argument and thought about it myself before you mentioned it. I don't think it should prevent us from using it, though. It is a feature of the Python programming language and anybody who's confused by it can probably look it up and figure out what it does eventually.